src/ImpactEvaluator.sol

```diff
@@ -134,18 +136,49 @@ contract ImpactEvaluator is AccessControl {
         uint64[] memory scores
     ) private {
         for (uint i = 0; i < addresses.length; i++) {
-            address payable addr = addresses[i];
-            if (addr == 0x000000000000000000000000000000000000dEaD) {
-                // TODO: Shall we emit an event here too?
-                continue;
-            }
-            uint256 amount = (scores[i] * roundReward) / MAX_SCORE;
-            if (addr.send(amount)) {
-                emit Transfer(addr, amount);
-            } else {
-                emit TransferFailed(addr, amount);
-            }
+            balances[addresses[i]] += (scores[i] * roundReward) / MAX_SCORE;
+        }
+    }
+
+    function balanceOf(address account) public view returns (uint) {
+        return balances[account];
+    }
+
+    function _withdraw(
+        address account,
+        address payable target,
+        uint value
+    ) private {
+        require(balances[account] >= value, "Insufficient balance");
+        balances[account] -= value;
+        if (balances[account] == 0) {
+            delete balances[account];
         }
+        require(target.send(value), "Withdrawal failed");
+    }
+
+    function withdraw(address payable target, uint value) public {
+        _withdraw(msg.sender, target, value);
+        emit Withdrawal(msg.sender, target, value);
```

Comment:
What is this event suitable for? If case you are going to use it to show "meridian withdrawals" in the Station UI:
- Isn't this information redundant? I mean the ledger should list the `target.send(value)` transaction, so we can already find them even without a new event.
- We have been evolving the smart contract over time, meaning that at different points in time, the Station has to look for events emitted by different smart contract addresses. How is this going to affect the implementation of function looking for reward withdrawals?

Reply:
> What is this event suitable for? If case you are going to use it to show "meridian withdrawals" in the Station UI:

Yes that is the goal

> Isn't this information redundant? I mean the ledger should list the target.send(value) transaction, so we can already find them even without a new event.

If we query contract transactions, we will find _all_ transactions, but can't filter by a specific account. We will see that the contract sent funds to `target`, but not which account it came from.

> We have been evolving the smart contract over time, meaning that at different points in time, the Station has to look for events emitted by different smart contract addresses. How is this going to affect the implementation of function looking for reward withdrawals?

This is a great point. I think Station will have to query transactions from all smart contracts it had been configured with. I don't see a particular issue with that. Do you?

Reply:
There is also another important reason for this change:

- Querying transactions by from/to address isn't supported by json-rpc endpoints. Therefore, we need to use block explorers
- Block explorers won't be available for IPC (initially, uncertain when one will be available)
- Therefore, relying on block explorers blocks our IPC adoption
- Querying the smart contract instead is a simple way to solve this

Reply:
Makes sense, thanks for explaining!

I agree with your conclusions.

> I think Station will have to query transactions from all smart contracts it had been configured with. I don't see a particular issue with that. Do you?

Not really, this sounds like a good plan to me 👍🏻